polyply/src/backmap.py

```diff
@@ -11,15 +11,128 @@
+               if attr[0] in molecule.nodes[node_b]:
+                  if molecule.nodes[node_b][attr[0]] == value[0]:
+                      edges.append(edge)
+
+    return edges
+
+
+def orient_template(meta_molecule, current_node, template, built_nodes):
+
+    # 1. find neighbours at meta_mol level
+    neighbours = nx.all_neighbors(meta_molecule, current_node)
+
+    # 2. find connecting atoms at low-res level
+    edges = []
+    for resid in neighbours:
+        edge = find_edges(meta_molecule.molecule,
+                            ("resid", "resid"),
+                            (resid+1, current_node+1))
+        edges += edge
+
+    # 3. build coordinate system
+    ref_coords = np.zeros((3, len(edges)))
+    opt_coords = np.zeros((3, len(edges)))
+    for ndx, edge in enumerate(edges):
+        atom_a, atom_b = edge
+        resid_a = meta_molecule.molecule.nodes[atom_a]["resid"] - 1
+        resid_b = meta_molecule.molecule.nodes[atom_b]["resid"] - 1
+        if resid_a in built_nodes or resid_b in built_nodes:
+            if resid_a == current_node:
+                 atom_name = meta_molecule.molecule.nodes[atom_a]["atomname"]
+                 aa_node = atom_b
+            else:
+                 atom_name = meta_molecule.molecule.nodes[atom_b]["atomname"]
+                 aa_node = atom_a
+
+            opt_coords[:, ndx] = template[atom_name]
+            ref_coords[:, ndx] = meta_molecule.molecule.nodes[aa_node]["position"] - meta_molecule.nodes[current_node]["position"]
+
+        else:
+            if resid_a == current_node:
+                 atom_name = meta_molecule.molecule.nodes[atom_a]["atomname"]
+                 cg_node = resid_b
+            else:
+                 atom_name = meta_molecule.molecule.nodes[atom_b]["atomname"]
+                 cg_node = resid_a
+
+            opt_coords[:, ndx] = template[atom_name]
+            ref_coords[:, ndx] = meta_molecule.nodes[cg_node]["position"] - meta_molecule.nodes[current_node]["position"]
+
+
+    # 4. minimize distance to meta_mol neighbours
+    def target_function(angles):
+        rotated = rotate_xyz(opt_coords, angles[0], angles[1], angles[2])
+        diff = rotated - ref_coords
+        score = np.sum(np.sqrt(np.sum(diff*diff, axis=0)))
+        return score
+
+    angles = np.array([10.0, -10.0, 5.0])
+    opt_results = scipy.optimize.minimize(target_function, angles, method='L-BFGS-B',
+                                              options={'ftol':0.000001, 'maxiter': 400})
+    # 5. rotate template
+    template_arr = np.zeros((3,len(template)))
+    key_to_ndx = {}
+    for ndx, key in enumerate(template.keys()):
+        template_arr[:, ndx] = template[key]
+        key_to_ndx.update({key:ndx})
+
+    angles = opt_results['x']
+    template_rotated_arr = rotate_xyz(template_arr, angles[0], angles[1], angles[2])
+    rotated = rotate_xyz(opt_coords, angles[0], angles[1], angles[2])
+    diff = rotated - ref_coords
```

Comment:
I'm not sure I understand the difference between the two rotations here. And you don't use diff.

Reply:
some sort of debug leftover I suppose